apis/console/src/lib.rs

```diff
@@ -66,6 +74,44 @@ impl<S: Syscalls, C: platform::allow_ro::Config + platform::subscribe::Config> C
         })
     }
 
+    /// Reads bytes
+    /// Reads from the device and writes to `buf`, starting from index 0.
+    /// No special guarantees about when the read stops.
+    /// Returns count of bytes written to `buf`.
+    pub fn read(buf: &mut [u8]) -> (usize, Result<(), ErrorCode>) {
+        let called = core::cell::Cell::new(Option::<(u32, u32)>::None);
+        let mut bytes_received = 0;
+        let r = share::scope::<
+            (
+                AllowRw<_, DRIVER_NUM, { allow_rw::READ }>,
+                Subscribe<_, DRIVER_NUM, { subscribe::READ }>,
+            ),
+            _,
+            _,
+        >(|handle| {
+            let (allow_rw, subscribe) = handle.split();
+            let len = buf.len();
+            S::allow_rw::<C, DRIVER_NUM, { allow_rw::READ }>(allow_rw, buf)?;
+            S::subscribe::<_, _, C, DRIVER_NUM, { subscribe::READ }>(subscribe, &called)?;
+
+            // When this fails, `called` is guaranteed unmodified,
+            // because upcalls are never processed within a `command` call.
```

Comment:
```suggestion
            // When this fails, `called` is guaranteed unmodified,
            // because upcalls are never processed until we call `yield`.
```

I think this is a slightly more accurate description? It seems the key here is that the appdata passed to subscribe is only modified in the libtock-rs upcall handler. @jrvanwhy is this correct?

Reply:
Yes, that is correct.